src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -679,9 +680,21 @@ namespace cryptonote
     r = m_mempool.init(max_txpool_weight, m_nettype == FAKECHAIN);
     CHECK_AND_ASSERT_MES(r, false, "Failed to initialize memory pool");
 
+    uint64_t session_start_height = m_blockchain_storage.get_current_blockchain_height();
+    uint8_t hardfork_version_start = m_blockchain_storage.get_hard_fork_version(session_start_height);
+    uint8_t hardfork_version_current = m_blockchain_storage.get_current_hard_fork_version();
```

Comment:
Will these two values ever be different at this point? This is `core::init`, the node hasn't started syncing at this point and is only loading existing blocks that it synced during last run.

Reply:
The discussion mentions that it can change if the node itself is updated in between runs, but it needs to be tested. So far it's not obvious that these two versions will be different even in this case.

Reply:
`heights` contains the HFs and `current_for_index` is the index of our current HF in the list. 

The case we want to verify is when we shutdown at boundary, and start up in next hf.

Reply:
What if someone shuts down a few heights above the boundary (they were late to the fork) and restarts with the new version?

Reply:
They would be on an invalid chain, I have thought about this and I don't think there is a good solution to this while we support voting for hardforks.

You would need to check all past HFs, or at least the ones outside the checkpoints and pop blocks if any blocks have incorrect major versions.

This is a separate issue though IMO

Reply:
@SChernykh I was wrong, we do handle this here:

https://github.com/monero-project/monero/blob/caa62bc9ea1c5f2ffe3ffa440ad230e1de509bfd/src/cryptonote_core/blockchain.cpp#L391-L429

The way we handle this though pretty much removes the possibility of using the voting for hardforks feature. We pop blocks until the minimum possible height of the HF, not the height it should activate according to votes, so even if the fork isn't active we will pop blocks.